lib/auth/bot.go

```diff
@@ -0,0 +1,286 @@
+
+	// Ensure existing resources don't already exist.
+	_, err := s.GetRole(ctx, resourceName)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if roleExists := (err == nil); roleExists {
+		return nil, trace.AlreadyExists("cannot add bot: role %q already exists", resourceName)
+	}
+
+	_, err = s.GetUser(resourceName, false)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if userExists := (err == nil); userExists {
+		return nil, trace.AlreadyExists("cannot add bot: user %q already exists", resourceName)
+	}
+
+	// Create the resources.
+	if err := createBotRole(ctx, s, req.Name, resourceName, req.Roles); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := createBotUser(ctx, s, req.Name, resourceName); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ttl := time.Duration(req.TTL)
+	if ttl == 0 {
+		ttl = defaults.DefaultBotJoinTTL
+	}
+
+	token, err := s.CreateBotJoinToken(ctx, CreateUserTokenRequest{
+		Name: resourceName,
+		TTL:  ttl,
+		Type: UserTokenTypeBot,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &proto.CreateBotResponse{
+		TokenID:  token.GetName(),
+		UserName: resourceName,
+		RoleName: resourceName,
+		TokenTTL: proto.Duration(ttl),
+	}, nil
+}
+
+// deleteBotUser removes an existing bot user, ensuring that it has bot labels
+// matching the bot before deleting anything.
+func (s *Server) deleteBotUser(ctx context.Context, botName, resourceName string) error {
+	user, err := s.GetUser(resourceName, false)
+	if err != nil {
+		err = trace.WrapWithMessage(err, "could not fetch expected bot user %s", resourceName)
+	} else {
+		label, ok := user.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete user %s that is missing label %s", resourceName, types.BotLabel)
+		} else if label != botName {
+			err = trace.Errorf("will not delete user %s with mismatched label %s = %s", resourceName, types.BotLabel, label)
+		} else {
+			err = s.DeleteUser(ctx, resourceName)
+		}
+	}
+
+	return err
+}
+
+// deleteBotRole removes an existing bot role, ensuring that it has bot labels
+// matching the bot before deleting anything.
+func (s *Server) deleteBotRole(ctx context.Context, botName, resourceName string) error {
+	role, err := s.GetRole(ctx, resourceName)
+	if err != nil {
+		err = trace.WrapWithMessage(err, "could not fetch expected bot role %s", resourceName)
+	} else {
+		label, ok := role.GetMetadata().Labels[types.BotLabel]
+		if !ok {
+			err = trace.Errorf("will not delete role %s that is missing label %s", resourceName, types.BotLabel)
+		} else if label != botName {
+			err = trace.Errorf("will not delete role %s with mismatched label %s = %s", resourceName, types.BotLabel, label)
+		} else {
+			err = s.DeleteRole(ctx, resourceName)
+		}
+	}
+
+	return err
+}
+
+func (s *Server) deleteBot(ctx context.Context, botName string) error {
+	// TODO:
+	// remove any locks for the bot's impersonator role?
```

Comment:
Do we have a separate issue where we're tracking locking of bots?

Reply:
I've mostly resolved bot locking in https://github.com/gravitational/teleport/pull/10098, I think this is the only remaining open question. I'm inclined to not delete any locks automatically since that would be new (and maybe unexpected) behavior since I don't think we delete locks when e.g. removing a user or role today.

If you agree I can just remove the TODO.

Reply:
I've replaced the TODO here with a note explaining why locks aren't currently deleted. Let me know if you'd prefer to see a behavior change instead!